File: app/src/main/java/com/example/buildmypc/ui/parts/parts/Case.java
package com.example.buildmypc.ui.parts.parts;

import android.os.Parcel;
import android.os.Parcelable;

import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;

public class Case extends Part implements Parcelable {
	public static final Creator<Case> CREATOR = new Creator<Case>() {
		@NotNull
		@Contract("_ -> new")
		@Override
		public Case createFromParcel(Parcel in) {
			return new Case(in);
		}

		@NotNull
		@Contract(value = "_ -> new", pure = true)
		@Override
		public Case[] newArray(int size) {
			return new Case[size];
		}
	};
	//hello
	private String color;
	private ArrayList<String> dimensionsMm; // 1-d int arraylist with 3 entries,
	private ArrayList<CountedString> expansionSlots;
	private ArrayList<String> supportedFrontUSBs;
	private ArrayList<CountedString> internalDriveBays;
	private ArrayList<String> maxGPULength;
	private ArrayList<String> mbFormFactor;
	private String sidePanel;
	private String type;
	private ArrayList<String> volume; // in liters because screw bri'ish units
	private boolean psuShroud;

	public Case(String model, String manufacturer) {
		super(model, manufacturer);
	}

	public Case(String model, String manufacturer, String color, ArrayList<String> dimensionsMm, ArrayList<CountedString> expansionSlots, ArrayList<String> supportedFrontUSBs, ArrayList<CountedString> internalDriveBays, ArrayList<String> maxGPULength, ArrayList<String> mbFormFactor, String sidePanel, String type, ArrayList<String> volume, boolean psuShroud) {
		super(model, manufacturer);
		this.color = color;
		this.dimensionsMm = dimensionsMm;
		this.expansionSlots = expansionSlots;
		this.supportedFrontUSBs = supportedFrontUSBs;
		this.internalDriveBays = internalDriveBays;
		this.maxGPULength = maxGPULength;
		this.mbFormFactor = mbFormFactor;
		this.sidePanel = sidePanel;
		this.type = type;
		this.volume = volume;
		this.psuShroud = psuShroud;
	}

	protected Case(@NotNull Parcel in) {
		super(in.readString(), in.readString());
		color = in.readString();
		dimensionsMm = in.createStringArrayList();
		expansionSlots = in.createTypedArrayList(CountedString.CREATOR);
		supportedFrontUSBs = in.createStringArrayList();
		internalDriveBays = in.createTypedArrayList(CountedString.CREATOR);
		maxGPULength = in.createStringArrayList();
		mbFormFactor = in.createStringArrayList();
		sidePanel = in.readString();
		type = in.readString();
		volume = in.createStringArrayList();
		psuShroud = in.readBoolean();
	}

	public Case(String model, String manufacturer, String color) {
		this(model, manufacturer, color, null, null, null, null, null, null, null, null, null, false);
	}

	@Override
	public void writeToParcel(@NotNull Parcel dest, int flags) {
		dest.writeString(super.getModel());
		dest.writeString(super.getManufacturer());
		dest.writeString(color);
		dest.writeStringList(dimensionsMm);
		dest.writeTypedList(expansionSlots);
		dest.writeStringList(supportedFrontUSBs);
		dest.writeTypedList(internalDriveBays);
		dest.writeStringList(maxGPULength);
		dest.writeStringList(mbFormFactor);
		dest.writeString(sidePanel);
		dest.writeString(type);
		dest.writeStringList(volume);
		dest.writeBoolean(psuShroud);
	}

	public String getColor() {
		return color;
	}

	public void setColor(String color) {
		this.color = color;
	}

	public ArrayList<String> getDimensionsMm() {
		return dimensionsMm;
	}

	public void setDimensionsMm(ArrayList<String> dimensionsMm) {
		this.dimensionsMm = dimensionsMm;
	}

	public ArrayList<CountedString> getExpansionSlots() {
		return expansionSlots;
	}

	public void setExpansionSlots(ArrayList<CountedString> expansionSlots) {
		this.expansionSlots = expansionSlots;
	}

	public ArrayList<String> getSupportedFrontUSBs() {
		return supportedFrontUSBs;
	}

	public void setSupportedFrontUSBs(ArrayList<String> supportedFrontUSBs) {
		this.supportedFrontUSBs = supportedFrontUSBs;
	}

	public ArrayList<CountedString> getInternalDriveBays() {
		return internalDriveBays;
	}

	public void setInternalDriveBays(ArrayList<CountedString> internalDriveBays) {
		this.internalDriveBays = internalDriveBays;
	}

	public ArrayList<String> getMaxGPULength() {
		return maxGPULength;
	}

	public void setMaxGPULength(ArrayList<String> maxGPULength) {
		this.maxGPULength = maxGPULength;
	}

	public ArrayList<String> getMbFormFactor() {
		return mbFormFactor;
	}

	public void setMbFormFactor(ArrayList<String> mbFormFactor) {
		this.mbFormFactor = mbFormFactor;
	}

	public String getSidePanel() {
		return sidePanel;
	}

	public void setSidePanel(String sidePanel) {
		this.sidePanel = sidePanel;
	}

	public String getType() {
		return type;
	}

	public void setType(String type) {
		this.type = type;
	}

	public ArrayList<String> getVolume() {
		return volume;
	}

	public void setVolume(ArrayList<String> volume) {
		this.volume = volume;
	}

	public boolean isPsuShroud() {
		return psuShroud;
	}

	public void setPsuShroud(boolean psuShroud) {
		this.psuShroud = psuShroud;
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (o == null || getClass() != o.getClass()) return false;
		if (!super.equals(o)) return false;

		Case aCase = (Case) o;

		if (isPsuShroud() != aCase.isPsuShroud()) return false;
		if (getColor() != null ? !getColor().equals(aCase.getColor()) : aCase.getColor() != null)
			return false;
		if (getDimensionsMm() != null ? !getDimensionsMm().equals(aCase.getDimensionsMm()) : aCase.getDimensionsMm() != null)
			return false;
		if (getExpansionSlots() != null ? !getExpansionSlots().equals(aCase.getExpansionSlots()) : aCase.getExpansionSlots() != null)
			return false;
		if (getSupportedFrontUSBs() != null ? !getSupportedFrontUSBs().equals(aCase.getSupportedFrontUSBs()) : aCase.getSupportedFrontUSBs() != null)
			return false;
		if (getInternalDriveBays() != null ? !getInternalDriveBays().equals(aCase.getInternalDriveBays()) : aCase.getInternalDriveBays() != null)
			return false;
		if (getMaxGPULength() != null ? !getMaxGPULength().equals(aCase.getMaxGPULength()) : aCase.getMaxGPULength() != null)
			return false;
		if (getMbFormFactor() != null ? !getMbFormFactor().equals(aCase.getMbFormFactor()) : aCase.getMbFormFactor() != null)
			return false;
		if (getSidePanel() != null ? !getSidePanel().equals(aCase.getSidePanel()) : aCase.getSidePanel() != null)
			return false;
		if (getType() != null ? !getType().equals(aCase.getType()) : aCase.getType() != null)
			return false;
		return getVolume() != null ? getVolume().equals(aCase.getVolume()) : aCase.getVolume() == null;
	}

	@Override
	public int hashCode() {
		int result = super.hashCode();
		result = 31 * result + (getColor() != null ? getColor().hashCode() : 0);
		result = 31 * result + (getDimensionsMm() != null ? getDimensionsMm().hashCode() : 0);
		result = 31 * result + (getExpansionSlots() != null ? getExpansionSlots().hashCode() : 0);
		result = 31 * result + (getSupportedFrontUSBs() != null ? getSupportedFrontUSBs().hashCode() : 0);
		result = 31 * result + (getInternalDriveBays() != null ? getInternalDriveBays().hashCode() : 0);
		result = 31 * result + (getMaxGPULength() != null ? getMaxGPULength().hashCode() : 0);
		result = 31 * result + (getMbFormFactor() != null ? getMbFormFactor().hashCode() : 0);
		result = 31 * result + (getSidePanel() != null ? getSidePanel().hashCode() : 0);
		result = 31 * result + (getType() != null ? getType().hashCode() : 0);
		result = 31 * result + (getVolume() != null ? getVolume().hashCode() : 0);
		result = 31 * result + (isPsuShroud() ? 1 : 0);
		return result;
	}

	@NotNull
	@Override
	public String toString() {
		return getColor() + " " + super.toString() + " " + getType();
		/*  White NZXT H510 ATX Mid Tower
			Black NZXT H510 ATX Mid Tower
			Black Corsair 4000D Airflow ATX Mid Tower
			Black Corsair 275R Airflow AIX Mid Tower
			Black Lian Li PC-O11DX ATX Full Tower */

	}

	@Override
	public int getParamCount() {
		return 13;
	}
}

File: app/src/main/java/com/example/buildmypc/ui/parts/parts/Case.kt
package com.example.buildmypc.ui.parts.parts

import android.os.Parcel
import android.os.Parcelable
import android.os.Parcelable.Creator

open class Case : Part, Parcelable {
	//hello
	var color: String? = null
	var dimensionsMm // 1-d int arraylist with 3 entries,
			: ArrayList<String>? = null
	var expansionSlots: ArrayList<CountedString?>? = null
	var supportedFrontUSBs: ArrayList<String>? = null
	var internalDriveBays: ArrayList<CountedString?>? = null
	var maxGPULength: ArrayList<String>? = null
	var mbFormFactor // motherboard form factor
			: ArrayList<String>? = null
	var sidePanel: String? = null
	var type: String? = null
	var volume // in liters because screw bri'ish units
			: ArrayList<String>? = null
	var isPsuShroud = false

	constructor(model: String?, manufacturer: String?) : super(model, manufacturer) {
		this.color = null
		this.dimensionsMm = null
		this.expansionSlots = null
		this.supportedFrontUSBs = null
		this.internalDriveBays = null
		this.maxGPULength = null
		this.mbFormFactor = null
		this.sidePanel = null
		this.type = null
		this.volume = null
		this.isPsuShroud = false
	}

	@JvmOverloads
	constructor(
		model: String?,
		manufacturer: String?,
		color: String?,
		dimensionsMm: ArrayList<String>? = null,
		expansionSlots: ArrayList<CountedString?>? = null,
		supportedFrontUSBs: ArrayList<String>? = null,
		internalDriveBays: ArrayList<CountedString?>? = null,
		maxGPULength: ArrayList<String>? = null,
		mbFormFactor: ArrayList<String>? = null,
		sidePanel: String? = null,
		type: String? = null,
		volume: ArrayList<String>? = null,
		psuShroud: Boolean = false
	) : super(model, manufacturer) {
		this.color = color
		this.dimensionsMm = dimensionsMm
		this.expansionSlots = expansionSlots
		this.supportedFrontUSBs = supportedFrontUSBs
		this.internalDriveBays = internalDriveBays
		this.maxGPULength = maxGPULength
		this.mbFormFactor = mbFormFactor
		this.sidePanel = sidePanel
		this.type = type
		this.volume = volume
		this.isPsuShroud = psuShroud
	}

	protected constructor(`in`: Parcel) : super(`in`.readString(), `in`.readString()) {
		color = `in`.readString()
		dimensionsMm = `in`.createStringArrayList()
		expansionSlots = `in`.createTypedArrayList(CountedString.CREATOR)
		supportedFrontUSBs = `in`.createStringArrayList()
		internalDriveBays = `in`.createTypedArrayList(CountedString.CREATOR)
		maxGPULength = `in`.createStringArrayList()
		mbFormFactor = `in`.createStringArrayList()
		sidePanel = `in`.readString()
		type = `in`.readString()
		volume = `in`.createStringArrayList()
		isPsuShroud = `in`.readBoolean()
	}

	override fun writeToParcel(dest: Parcel, flags: Int) {
		dest.writeString(super.model)
		dest.writeString(super.manufacturer)
		dest.writeString(color)
		dest.writeStringList(dimensionsMm)
		dest.writeTypedList(expansionSlots)
		dest.writeStringList(supportedFrontUSBs)
		dest.writeTypedList(internalDriveBays)
		dest.writeStringList(maxGPULength)
		dest.writeStringList(mbFormFactor)
		dest.writeString(sidePanel)
		dest.writeString(type)
		dest.writeStringList(volume)
		dest.writeBoolean(isPsuShroud)
	}

	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (other == null || javaClass != other.javaClass) return false
		if (!super.equals(other)) return false
		val aCase = other as Case
		if (isPsuShroud != aCase.isPsuShroud) return false
		if (if (color != null) color != aCase.color else aCase.color != null) return false
		if (if (dimensionsMm != null) dimensionsMm != aCase.dimensionsMm else aCase.dimensionsMm != null) return false
		if (if (expansionSlots != null) expansionSlots != aCase.expansionSlots else aCase.expansionSlots != null) return false
		if (if (supportedFrontUSBs != null) supportedFrontUSBs != aCase.supportedFrontUSBs else aCase.supportedFrontUSBs != null) return false
		if (if (internalDriveBays != null) internalDriveBays != aCase.internalDriveBays else aCase.internalDriveBays != null) return false
		if (if (maxGPULength != null) maxGPULength != aCase.maxGPULength else aCase.maxGPULength != null) return false
		if (if (mbFormFactor != null) mbFormFactor != aCase.mbFormFactor else aCase.mbFormFactor != null) return false
		if (if (sidePanel != null) sidePanel != aCase.sidePanel else aCase.sidePanel != null) return false
		if (if (type != null) type != aCase.type else aCase.type != null) return false
		return if (volume != null) volume == aCase.volume else aCase.volume == null
	}

	override fun hashCode() =
		31 * (31 * (31 * (31 * (31 * (31 * (31 * (31 * (31 * (31 * (31 * super.hashCode() + if (color != null) color.hashCode() else 0) + if (dimensionsMm != null) dimensionsMm.hashCode() else 0) + if (expansionSlots != null) expansionSlots.hashCode() else 0) + if (supportedFrontUSBs != null) supportedFrontUSBs.hashCode() else 0) + if (internalDriveBays != null) internalDriveBays.hashCode() else 0) + if (maxGPULength != null) maxGPULength.hashCode() else 0) + if (mbFormFactor != null) mbFormFactor.hashCode() else 0) + if (sidePanel != null) sidePanel.hashCode() else 0) + if (type != null) type.hashCode() else 0) + if (volume != null) volume.hashCode() else 0) + if (isPsuShroud) 1 else 0

	override fun toString() = "$color ${super.toString()} $type"

	override val paramCount = 13

	companion object {
		@JvmField
		val CREATOR = object : Creator<Case> {
			override fun createFromParcel(`in`: Parcel) = Case(`in`)

			override fun newArray(size: Int) = arrayOfNulls<Case>(size)
		}
	}
}